src/Chainweb/OpenAPIValidation.hs

```diff
@@ -0,0 +1,70 @@
+{-# language TupleSections #-}
+{-# language ScopedTypeVariables #-}
+{-# language FlexibleContexts #-}
+{-# language OverloadedStrings #-}
+
+module Chainweb.OpenAPIValidation (mkValidationMiddleware) where
+
+import Control.Monad
+import Data.Aeson
+import qualified Data.ByteString.Char8 as BS8
+import qualified Data.ByteString.Lazy as BL
+import Data.Foldable
+import qualified Data.HashSet as HS
+import Data.IORef
+import Data.LogMessage (LogFunctionText)
+import qualified Data.Text.Encoding as T
+import qualified Data.Yaml as Yaml
+import qualified Network.HTTP.Client as HTTP
+import Network.Wai
+import qualified Network.Wai.Middleware.Validation as WV
+import System.LogLevel
+
+import Chainweb.Logger
+import Chainweb.Time
+import Chainweb.Utils
+import Chainweb.Version
+
+mkValidationMiddleware :: Logger logger => logger -> ChainwebVersion -> HTTP.Manager -> IO Middleware
+mkValidationMiddleware logger v mgr = do
+    (chainwebSpec, pactSpec) <- fetchOpenApiSpecs
+    apiCoverageRef <- newIORef $ WV.initialCoverageMap [chainwebSpec, pactSpec]
+    apiCoverageLogTimeRef <- newIORef =<< getCurrentTimeIntegral
+    return $
+        WV.mkValidator apiCoverageRef (WV.Log logValidationFailure (logApiCoverage apiCoverageLogTimeRef)) $ \path -> asum
+            [ case BS8.split '/' path of
+                ("" : "chainweb" : "0.0" : rawVersion : "chain" : rawChainId : "pact" : "api" : "v1" : rest) -> do
+                    findPact pactSpec rawVersion rawChainId rest
+                ("" : "chainweb" : "0.0" : rawVersion : "chain" : rawChainId : "pact" : rest) -> do
+                    findPact pactSpec rawVersion rawChainId rest
+                _ -> Nothing
+            , (,chainwebSpec) <$> BS8.stripPrefix (T.encodeUtf8 $ "/chainweb/0.0/" <> chainwebVersionToText v) path
+            , Just (path,chainwebSpec)
+            ]
+    where
+    findPact pactSpec rawVersion rawChainId rest = do
+        reqVersion <- chainwebVersionFromText (T.decodeUtf8 rawVersion)
+        guard (reqVersion == v)
+        reqChainId <- chainIdFromText (T.decodeUtf8 rawChainId)
+        guard (HS.member reqChainId (chainIds v))
+        return (BS8.intercalate "/" ("":rest), pactSpec)
+
+    fetchOpenApiSpecs = do
+        let chainwebUri = "https://raw.githubusercontent.com/kadena-io/chainweb-openapi/main/chainweb.openapi.yaml"
```

Comment:
are these going to change at all? or are these the standard locations for these openapi definitions?

Reply:
This is where we can find the latest spec, ideally we could configure it somehow, maybe with an envvar.